### ISSUE TITLE
[Feature request] Add a FediLinks protocol handler

### ISSUE THREAD

**Issue by SoniEx2** [omarroth/invidious#3964]
**Is your feature request related to a problem? Please describe.**

Would like to be able to share youtube links without pointing ppl at youtube specifically. Ppl should just be able to use Invidious without messing with editing URLs. 

**Describe the solution you'd like**

~~We made Fedi-To to direct ppl to the websites *they* wanna use. So for example we could wrap a `web+video://youtube.com/watch?v=whatever` link in a Fedi-To link, have it open in youtube for your average user, but have it open in invidious for your invidious user. This only requires the website to provide a protocol handler endpoint, similar to one used for registerProtocolHandler.~~

We made [The Fedi Links Project](https://fedilinks.org) to improve mastodon and promote Atom feeds and other decentralized/distributed projects, and we're now up to 3 mastodon clients that implement it.

**Describe alternatives you've considered**

~~You could just use registerProtocolHandler directly but nobody uses that for a whole host of reasons, including the fact that you can't link to a registerProtocolHandler on some services. Meanwhile Fedi-To wrapped links are plain https and just work everywhere.~~

Not really relevant anymore.

**Additional context**

Nothing else really relevant.

~~Oh right: https://fedi-to.net/~~~

**Comment by Yetangitu** [comment]
I don't think this is the right way to go at decentralising the 'net as it actually adds a centralised step - the fedi-to dereferencer - to opening any links. Decisions like this should be made locally, anywhere from LAN-level to the application used to access content, without referencing any intermediates which could track whatever the user is doing.

Then there is the (ab)use of the protocol field to indicate content type, here I'd suggest simply using a parameter - something like what the Content-Type header is supposed to be used for but used as a parameter instead of a header. Those who have the required extension/add-on/content script installed will be referred to their proxy or application of preference while others will just use the original source. For Youtube this could look like:

`https://youtube.com/watch?v=whatever&x-content-type=web+video`

Of course the name of the parameter should be chosen so that collisions are unlikely, `x-content-type` is probably not the best choice.

Then there is the question of what this all has to do with federation as implied by the 'fedi-to' moniker. I understand that federation is all the rage at the moment with the Twitter-brouhaha, MetaFacebook's coming entry to this field and Reddit's antics but the use of content proxies like Invidious/Nitter/libreddit/etc is not related to federation in any way. Maybe 'link-to' would be a better name, or 'proxy-to', or 'gimme' or something like the 'sensible-browser/editor/pager/terminal' from Debian's _sensible-utils_ package would be better, e.g. sensible-viewer.

BTW, that "made by a queer trans person" reference in the fedi-to.net site is quite divisive, it may act like a rainbow flag to some but acts as a red flag to many others. May I suggest keeping sexual orientation (etc.) out of unrelated communications?

**Comment by SoniEx2** [comment]
If you don't like the fedi-to website being a website, make it a built-in OS/browser feature. Not like it really matters; the web is centralized, it's all controlled by the WHATWG anyway. Pick your poison.

**Comment by markozajc** [comment]
I'm sure anyone interested in using Invidious and similar services uses or at least knows about redirection extensions such as [Privacy Redirect](https://addons.mozilla.org/en-US/firefox/addon/privacy-redirect/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search), [Redirector](https://addons.mozilla.org/en-US/firefox/addon/redirector/), or similar - there's no need for "messing with editing URLs" if you want to use Invidious.

I'm also having trouble discerning what feature exactly is this requesting: in the title you suggest that Invidious starts registering a protocol handler (for `web+video` if I understood correctly), but in the body you explain how "nobody uses that" and pitch your own website that can support redirection without changes to Invidious.

**Comment by SoniEx2** [comment]
the idea is to familiarize users with protocol handlers and set them free from traditional platforms.

the goal is to EEE the traditional platforms.

**Comment by markozajc** [comment]
Okay, but what's the Invidious feature being requested exactly?

**Comment by SoniEx2** [comment]
a protocol handler endpoint.

we're not asking Invidious to actually call registerProtocolHandler, but to have an endpoint for a future registerProtocolHandler-equivalent.

**Comment by SamantazFox** [comment]
What's the benefit if this method over sharing youtube URLs and using a redirector extension (or alternatively sharing redirect.invidious.io URLs and selecting an instance)?

**Comment by SoniEx2** [comment]
wait what's redirect.invidious.io

**Comment by SoniEx2** [comment]
okay that's pretty cool. so the idea was to standardize an API for that kind of stuff, that also makes it user-overridable. so you could have e.g. `web+invidious://redirect.invidious.io/...` and have it open in `redirect.invidious.io` by default, but then have user-defined protocol handlers that are preferred over it, without having to chase it in the big-ass list.

(but we're still working on that "standardize" part.)

**Comment by SamantazFox** [comment]
One thing I don't understand, is how do you handle these `web+://` schemes?

> So for example we could wrap a `web+video://youtube.com/watch?v=whatever` link in a Fedi-To link, have it open in youtube for your average user, but have it open in invidious for your invidious user. This only requires the website to provide a protocol handler endpoint, similar to one used for registerProtocolHandler.

Who "registers" the protocol handler? To whom?
Do all invidious instances have to register that handler?
If yes, what do you do when you have 90 handlers for `web+video://`? Do you redirect randomly, or are you storing user data to know which instance someone prefers?

How is sharing `web+video://youtube.com/watch...` different from sharing a regular youtube URL and letting the user install (or not) an extension to redirect them?

---

An other thing I don't understand, is who hosts this API you're talking about?
If that's centralized on your server, what if it goes down?

**Comment by SoniEx2** [comment]
oh no, we don't store user data. user stores user data. (it's called cookies, they're pretty great if you use them for their intended purpose, instead of tracking or profiling.)

the long-term goal is to not have a centralized server, but yes the current approach is to just throw a centralized server at the problem. or to use one of the flows described in the [Internet-Draft](https://datatracker.ietf.org/doc/draft-soni-protocol-handler-well-known-uri/).

how would an extension work? is invidious detectable by an extension? are there browser APIs for detecting whether a website does a thing (like "is an alternative youtube frontend") or does that rely on guesswork?

**Comment by SamantazFox** [comment]
From what I can grasp of this internet draft, each website have to implement something (a sort of popup?) in which the user will input a domain, and then get redirected if that domain supports that protocol handler, right? Isn't it a bit cumbersome? How is this faster than copy/pasting the URL into your home instance's search box (in the case of mastodon or piped, e.g)?

> how would an extension work? is invidious detectable by an extension? are there browser APIs for detecting whether a website does a thing (like "is an alternative youtube frontend") or does that rely on guesswork?

The extension I use (libredirect) works with a list of know instances URL per service (e.g youtube). But some others (e.g redirector) works with pure URL regex. List of instances can be gathered from online sources (like our official list) or managed locally.

**Comment by SoniEx2** [comment]
the Internet-Draft defines no particular interaction flow with the endpoint - the examples are mostly informative. but yes, that is exactly what the examples demonstrate. a perhaps more exciting interaction flow is provided by the `fedicraft` minecraft mod, but that is not supported by web browsers at this point.

the extension-based workaround sounds... well, it sounds *fine*. it does require installing an extension which is often not possible, so that makes it not great. but it's fine, if you can use it. but is that good enough? not everyone might agree that it is.

**Comment by markozajc** [comment]
Can you explain in detail and without vagueness how this would work for Invidious and which changes need to be made?

**Comment by SoniEx2** [comment]
invidious would define an URI format: `web+invidious://some.invidious.instance.example/watch?v=video_id`

invidious would then implement `GET /.well-known/protocol-handler?target=web+invidious://some.invidious.instance.example/watch?v=video_id` to redirect to `Location: /watch?v=video_id`

and that's it.
